students/kegan/session10/MailroomTools.py

```diff
@@ -0,0 +1,489 @@
+        Thanks donor for all donations if all_donations is True.
+        Args:
+            all_donations (bool) : thank donor for all donations
+        Returns:
+            str : thank you message for donor
+        """
+        num = self.num if all_donations else 1
+        first = self.donations[:-1] if all_donations else []
+        rest = self.donations[-1]
+        substrings = {
+            'donor': self.name,
+            's': 's' if num > 1 else '',
+            'and': ' and ' if num > 1 else '',
+            'first': ', '.join(['${:,.2f}'.format(d) for d in first]),
+            'rest': '${:,.2f}'.format(rest),
+            'totalling':
+                '' if num < 2 else
+                ', totalling {}{},'.format(
+                    'an incredible ' if self.total > 500
+                    else '', '${:,.2f}'.format(self.total))}
+        message = \
+            'Dear {donor},\nThank you for your generous gift{s} of ' +\
+            '{first}{and}{rest}. Your donation{s}{totalling} will ' +\
+            'go towards feeding homeless kittens in Seattle. ' +\
+            'From the bottom of our hearts, we at Miuvenile Care thank you.' +\
+            '\n\nRegards,\nBungelina Bigglesnorf\nChairwoman, Miuvenile Care'
+        return message.format(**substrings)
+
+    def add(self, *donations):
+        """ Adds passed donations to this donor.
+        Args:
+            donations (args) : donations as arguments
+        """
+        donations = self.intake_donations(*donations)
+        self._donations.extend(donations)
+
+    def __str__(self):
+        """ Returns this donor as a string.
+        Returns:
+            str : donor as string
+        """
+        donations = ', '.join([
+            '${:,.2f}'.format(d) for d in self.donations])
+        return '{}: {}'.format(self.name, donations)
+
+    def __repr__(self):
+        """ Returns representation of this donor.
+        Returns:
+            str : representation of this donor
+        """
+        donations = ', '.join([
+            str(d) for d in self.donations])
+        return 'Donor("{}", {})'.format(self.name, donations)
+
+    def __eq__(self, other):
+        """ Returns whether this donor is equal to other in
+        both name and all donations.
+        Args:
+            other (Donor) : donor to compare
+        Returns:
+            bool : True if other donor is equal to this donor
+        """
+        return (
+            self.name == other.name and
+            self.donations == other.donations)
+
+    def __lt__(self, other):
+        """ Returns whether this donor is less than
+        other. Compares name and donations.
+        Args:
+            other (Donor) : donor to compare
+        Returns:
+            bool : True if this donor is less than other
+        """
+        return (
+            self.name < other.name and
+            self.donations < other.donations)
+
+    def __contains__(self, donation):
```

Comment:
is this useful?

not bad practice, but there are a lot of magic methods, you generally don't implement ones that aren't useful.

Also, you have exposed a donations list, s if someone rally wanted to know if a particular donation amount was in there:

`amount in donor.donations`

would work, and not require them knowing anything "private" about the implementation.

Though you are right that if getting the donations list was expensive (Pulling for a DB, etc) than it might be good to have a __contains__ method that could do an efficient DB query.